### DISCUSSION TITLE
Method of generating code

### DISCUSSION THREAD

**Discussion by amake** [petitparser/dart-petitparser#179] [General]
I am looking into generating code for a package of mine, and am evaluating methods for doing this. 

I noticed that some of the code in this package is generated:
https://github.com/petitparser/dart-petitparser/tree/main/lib/src/parser/combinator/generated

I didn't see any mention of the tools or methods used. What are they, and how do you find them?

**Comment by amake** [comment]
Oops! I just found the generator:
https://github.com/petitparser/dart-petitparser/blob/main/bin/generate_sequence.dart

I guess that answers my question, although I'm curious if you considered using something like [source_gen](https://pub.dev/packages/source_gen).

**Comment by renggli** [comment]
Here are a bunch of generators in the same style from other packages I wrote:

- https://github.com/renggli/dart-more/blob/main/bin/generate_tuple.dart
- https://github.com/renggli/dart-more/blob/main/bin/generate_types.dart
- https://github.com/renggli/dart-more/blob/main/bin/generate_unicode.dart
- https://github.com/renggli/dart-rtlsdr/blob/main/bin/generate_bindings.dart

I don't think helper packages were around when I first wrote these code generators, and afterwards I just forked and adopted them as needed. Also [source_gen](https://pub.dev/packages/source_gen) and the upcoming [macros](https://dart.dev/language/macros) (which I am very much excited about) are more about extending and transforming existing code, in all my cases I just wanted to one-time generate some repetitive code in the package.

**Reply by amake** [reply]
Thank you for pointing me to this. I was halfway to making my own (bad, inefficient) version of your UnicodeCharMatcher, but now I don't have to :)
